empyrical/tests/test_stats.py

```diff
@@ -955,3 +971,159 @@ def test_cagr_noisy(self, returns, add_noise):
+        """
+        This returns a wrapper around the empyrical module, used a hook, so
+        tests perform, e.g. input conversions or return type checks, on each
+        call to an empyrical function. Each subclass can define this property
+        so that all the same tests are run but with different function inputs
+        or type checks.
+
+        This was done as part of enabling empyrical functions to work with
+        inputs of either pd.Series or np.ndarray, with the expectation that
+        they will return the same type as their input.
+
+        Returns
+        -------
+        empyrical
+
+        Notes
+        -----
+        Since some parameterized test parameters refer to attributes on the
+        real empyrical module at class body scope, this property must be
+        defined later in the body than those references. That way, the
+        attributes are looked up on the empyrical module, not this property.
+
+        e.g. empyrical.DAILY
+        """
+        return ReturnTypeEmpyricalProxy(self, (pd.Series, float))
+
+
+class TestStatsArrays(TestStats):
+    """
+    Tests pass np.ndarray inputs to empyrical and assert that outputs are of
+    type np.ndarray or float.
+
+    """
+    @property
+    def empyrical(self):
+        return PassArraysEmpyricalProxy(self, (np.ndarray, float))
+
+
+class TestStatsIntIndex(TestStats):
+    """
+    Tests pass int-indexed pd.Series inputs to empyrical and assert that
+    outputs are of type pd.Series or float.
+
+    This prevents a regression where we're indexing with ints into a pd.Series
+    to find the last item and get a KeyError when the series is int-indexed.
+
+    """
+    @property
+    def empyrical(self):
+        return ConvertPandasEmpyricalProxy(
+            self,
+            (pd.Series, float),
+            lambda obj: type(obj)(obj.values, index=np.arange(len(obj))),
+        )
+
+
+class ReturnTypeEmpyricalProxy(object):
+    """
+    A wrapper around the empyrical module which, on each function call, asserts
+    that the type of the return value is in a given set.
+
+    Calling an instance with kwargs will return a new copy with those
+    attributes overridden.
+
+    """
+    def __init__(self, test_case, return_types):
+        self._test_case = test_case
+        self._return_types = return_types
+
+    def __call__(self, **kwargs):
+        dupe = copy(self)
+
+        for k, v in iteritems(kwargs):
+            attr = '_' + k
+            if hasattr(dupe, attr):
+                setattr(dupe, attr, v)
+
+        return dupe
+
+    def __copy__(self):
+        newone = type(self).__new__(type(self))
+        newone.__dict__.update(self.__dict__)
+        return newone
+
+    def __getattr__(self, item):
+        return self._check_return_type(getattr(empyrical, item))
+
+    def _check_return_type(self, func):
+        @wraps(func)
+        def check_return_type(*args, **kwargs):
+            result = func(*args, **kwargs)
```

Comment:
Would be pretty cool to extend this to validate that the input data was not mutated (given that we also removed my paranoid `returns.copy()` in `_adjust_returns`).